File: Groak_iOS/Groak/Groak/RestaurantObjects/Dish.swift
//
//  Dish.swift
//  Groak
//
//  Created by Vishisht Tiwari on 11/18/19.
//  Copyright © 2019 Groak. All rights reserved.
//
//  This file contains classes for dish, dish extra and dish extra options

import Foundation
import Firebase

internal class Dish {
    var reference: DocumentReference?
    var name: String
    var imageLink: String
    var price: Double
    var shortInfo: String
    var dishDescription: String
    var ingredients: [String]
    var restaurantReference: DocumentReference?
    var available: Bool
    var restrictions: [String: String]
    var nutrition: [String: Double]
    var extras: [DishExtra]
    
    init() {
        reference = nil
        name = ""
        imageLink = ""
        price = -1
        shortInfo = ""
        dishDescription = ""
        ingredients = []
        restaurantReference = nil
        available = false
        restrictions = [:]
        nutrition = [:]
        extras = []
    }
    
    init(dish: [String: Any]) {
        reference = dish["reference"] as? DocumentReference
        name = dish["name"] as? String ?? ""
        imageLink = dish["image"] as? String ?? ""
        price = dish["price"] as? Double ?? -1
        shortInfo = dish["shortInfo"] as? String ?? ""
        dishDescription = dish["description"] as? String ?? ""
        ingredients = dish["ingredients"] as? [String] ?? []
        restaurantReference = dish["restaurantReference"] as? DocumentReference
        available = dish["available"] as? Bool ?? false
        restrictions = dish["restrictions"] as? [String: String] ?? [:]
        nutrition = dish["nutrition"] as? [String: Double] ?? [:]
        let tempExtras = dish["extras"] as? [[String:Any]] ?? []
        extras = []
        for extra in tempExtras {
            extras.append(DishExtra.init(dishExtra: extra))
        }
    }
    
    func restrictionsExist() -> Bool {
        if restrictions.count == 0 {
            return false
        }
        for (_, isit) in restrictions {
            if isit != "Not Sure" {
                return true
            }
        }
        return false
    }
    
    func nutritionExist() -> Bool {
        if nutrition.count == 0 {
            return false
        }
        for (_, value) in nutrition {
            if value > 0 {
                return true
            }
        }
        return false
    }
    
    func success() -> Bool {
        if (reference == nil) {
            return false;
        } else if (price < 0) {
            return false;
        } else if (name.count == 0) {
            return false;
        } else if (restaurantReference == nil) {
            return false;
        }
        
        return true;
    }
    
    var description: String {
        var str = "Dish Name: \(name)\n"
        str += "Dish Image Link: \(imageLink)\n"
        str += "Price: \(price)\n"
        str += "Short Info: \(shortInfo)\n"
        str += "Description: \(dishDescription)\n"
        str += "Ingredients:\n"
        for ingredient in ingredients {
            str += "\t\(ingredient)\n"
        }
        str += "Available: \(available)\n"
        str += "Restrictions:\n"
        for (restriction, isit) in restrictions {
            str += "\t\(restriction): \(isit)\n"
        }
        str += "Nutrition:\n"
        for (nut, value) in nutrition {
            str += "\t\(nut): \(value)\n"
        }
        str += "Extras:\n"
        for extra in extras {
            str += "\t\(extra.description):\n"
        }
        return str;
    }
}

internal class DishExtra {
    var title: String
    var multipleSelections: Bool
    var minOptionsSelect: Int
    var maxOptionsSelect: Int
    var options: [DishExtraOption]
    
    init(dishExtra: [String: Any]) {
        title = dishExtra["title"] as? String ?? ""
        multipleSelections = dishExtra["multipleSelections"] as? Bool ?? true
        minOptionsSelect = dishExtra["minOptionsSelect"] as? Int ?? -1
        maxOptionsSelect = dishExtra["maxOptionsSelect"] as? Int ?? -1
        let tempOptions = dishExtra["options"] as? [[String:Any]] ?? []
        options = []
        for option in tempOptions {
            options.append(DishExtraOption.init(dishExtraOption: option))
        }
    }
    
    init(cartDishExtra: CartDishExtra) {
        self.title = cartDishExtra.title
        self.options = []
        for option in cartDishExtra.options {
            self.options.append(DishExtraOption.init(title: option.title, price: option.price))
        }
        multipleSelections = true
        minOptionsSelect = -1
        maxOptionsSelect = -1
    }
    
    init(orderDishExtra: OrderDishExtra) {
        self.title = orderDishExtra.title
        self.options = []
        for option in orderDishExtra.options {
            self.options.append(DishExtraOption.init(title: option.title, price: option.price))
        }
        multipleSelections = true
        minOptionsSelect = -1
        maxOptionsSelect = -1
    }
    
    func success() -> Bool {
        if (title.count == 0) {
            return false;
        } else if (minOptionsSelect == -1) {
            return false;
        } else if (maxOptionsSelect == -1) {
            return false;
        }
        
        return true;
    }
    
    var description : String {
        var str = "Dish Extra Name: \(title)\n"
        str += "Dish Extra Multiple Selections : \(multipleSelections)\n"
        str += "Dish Extra Min Options Select: \(minOptionsSelect)\n"
        str += "Dish Extra Max Options Select: \(maxOptionsSelect)\n"
        str += "Dish Extra Options: \(options)\n"
        str += "Dish Extra Options :\n"
        for option in options {
            str += "\t\(option.description)\n"
        }
        
        return str;
    }
}

internal class DishExtraOption {
    var title: String
    var price: Double
    
    init(dishExtraOption: [String: Any]) {
        title = dishExtraOption["title"] as? String ?? ""
        price = dishExtraOption["price"] as? Double ?? -1
    }
    
    init(title: String, price: Double) {
        self.title = title
        self.price = price
    }
    
    func success() -> Bool {
        if (title.count == 0) {
            return false;
        } else if (price < 0) {
            return false;
        }
        
        return true;
    }
    
    var description : String {
        var str = "Dish Extra Option Title: \(title)\n"
        str += "Dish Extra Option Price: \(price)"
        
        return str;
    }
}

File: Groak_android/app/src/main/java/com/groak/groak/restaurantobject/dish/Dish.java
/**
 * Dish
 */
package com.groak.groak.restaurantobject.dish;

import com.google.firebase.firestore.DocumentReference;
import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.reflect.TypeToken;
import com.groak.groak.firebase.Firebase;
import com.groak.groak.restaurantobject.dish.dishextra.DishExtra;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class Dish {
    private DocumentReference reference;
    private String name;
    private String imageLink;
    private double price;
    private String shortInfo;
    private String description;
    private ArrayList<String> ingredients;
    private DocumentReference restaurantReference;
    private boolean available;
    private HashMap<String, String> restrictions;
    private HashMap<String, Object> nutrition;
    private ArrayList<DishExtra> extras;

    public Dish() {
        this.reference = null;
        this.name = "";
        this.imageLink = "";
        this.price = 0;
        this.shortInfo = "";
        this.description = "";
        this.ingredients = new ArrayList<>();
        this.restaurantReference = null;
        this.available = false;
        this.restrictions = new HashMap<>();
        this.nutrition = new HashMap<>();
        this.extras = new ArrayList<>();
    }

    public Dish(Map<String, Object> map) {
        this.reference = (DocumentReference)map.get("reference");
        this.name = (String)map.get("name");
        this.imageLink = (String)map.get("image");
        try {
            this.price = (double) map.get("price");
        } catch (ClassCastException e) {
            this.price = (long) map.get("price");
        }
        this.shortInfo = (String)map.get("shortInfo");
        this.description = (String)map.get("description");
        this.ingredients = (ArrayList<String>)map.get("ingredients");
        this.restaurantReference = (DocumentReference) map.get("restaurantReference");
        this.available = (boolean)map.get("available");
        this.restrictions = (HashMap<String, String>) map.get("restrictions");
        this.nutrition = (HashMap<String, Object>) map.get("nutrition");
        this.extras = new ArrayList<>();
        ArrayList<Map<String, Object>> tempExtras = (ArrayList<Map<String, Object>>)map.get("extras");
        for (Map<String, Object> extra: tempExtras) {
            this.extras.add(new DishExtra(extra));
        }
    }

    public Dish(JsonObject json) {
        Gson gson = new Gson();

        if (json.get("reference") != null)
            this.reference = Firebase.firebase.db.document(json.get("reference").getAsString());
        if (json.get("name") != null)
            this.name = json.get("name").getAsString();
        if (json.get("imageLink") != null)
            this.imageLink = json.get("imageLink").getAsString();
        if (json.get("price") != null)
            this.price = json.get("price").getAsDouble();
        if (json.get("shortInfo") != null)
            this.shortInfo = json.get("shortInfo").getAsString();
        if (json.get("description") != null)
            this.description = json.get("description").getAsString();
        if (json.get("ingredients") != null)
            this.ingredients = gson.fromJson(json.get("ingredients").getAsJsonArray(), new TypeToken<ArrayList<String>>(){}.getType());
        if (json.get("restaurantReference") != null)
            this.restaurantReference = Firebase.firebase.db.document(json.get("restaurantReference").getAsString());
        if (json.get("available") != null)
            this.available = json.get("available").getAsBoolean();
        if (json.get("restrictions") != null)
            this.restrictions = gson.fromJson(json.get("restrictions").getAsJsonObject(), new TypeToken<HashMap<String, String>>(){}.getType());
        if (json.get("nutrition") != null)
            this.nutrition = gson.fromJson(json.get("nutrition").getAsJsonObject(), new TypeToken<HashMap<String, Double>>(){}.getType());
        if (json.get("extras") != null)
            this.extras = gson.fromJson(json.get("extras").getAsJsonArray(), new TypeToken<ArrayList<DishExtra>>(){}.getType());
    }

    public DocumentReference getReference() {
        return reference;
    }
    public String getName() {
        return name;
    }
    public String getImageLink() {
        return imageLink;
    }
    public double getPrice() {
        return price;
    }
    public String getShortInfo() {
        return shortInfo;
    }
    public String getDescription() {
        return description;
    }
    public ArrayList<String> getIngredients() {
        return ingredients;
    }
    public DocumentReference getRestaurantReference() {
        return restaurantReference;
    }
    public boolean isAvailable() {
        return available;
    }
    public HashMap<String, String> getRestrictions() {
        return restrictions;
    }
    public HashMap<String, Object> getNutrition() {
        return nutrition;
    }
    public ArrayList<DishExtra> getExtras() {
        return extras;
    }

    public String toString() {
        String str = "";

        str += "Reference: " + reference + "\n";
        str += "Name: " + name + "\n";
        str += "Image Link: " + imageLink + "\n";
        str += "Price: " + price + "\n";
        str += "Short Info: " + shortInfo + "\n";
        str += "Description: " + description + "\n";
        str += "Ingredients: " + ingredients + "\n";
        str += "Available: " + available + "\n";
        str += "Restrictions: " + restrictions + "\n";
        str += "Nutrition: " + nutrition + "\n";
        str += "Extras: " + extras.toString() + "\n";

        return str;
    }
}